ToolBox/GoogleDownloader/GoogleDataDownloader.cs

```diff
@@ -44,21 +39,122 @@ public class GoogleDataDownloader : IDataDownloader
+        private enum ConvertMonths
+        {
+            Jan = 1,
+            Feb = 2,
+            Mar = 3,
+            Apr = 4,
+            May = 5,
+            Jun = 6,
+            Jul = 7,
+            Aug = 8,
+            Sep = 9,
+            Oct = 10,
+            Nov = 11,
+            Dec = 12
+        };
+
+        private IEnumerable<BaseData> GetDaily(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            var startdate = ((ConvertMonths)startUtc.Month).ToString()
+                + @"+" + startUtc.Day.ToString()
+                + @"%2C+" + startUtc.Year.ToString();
+            var enddate = ((ConvertMonths)endUtc.Month).ToString()
+                + @"+" + endUtc.Day.ToString()
+                + @"%2C+" + endUtc.Year.ToString();
+
+            // Create the Google formatted URL.
+            var url = string.Format(UrlPrototypeDaily, ConvertTicker(symbol), startdate, enddate);
+
+            // Download the data from Google.
+            string[] lines;
+            using (var client = new WebClient())
+            {
+                var data = client.DownloadString(url);
+                lines = data.Split('\n');
+            }
+
+            // first line is header
+            var currentLine = 1;
+
+            while (currentLine < lines.Length - 1)
+            {
+                // Format: Date,Open,High,Low,Close,Volume
+                var columns = lines[currentLine].Split(',');
+
+                // date format: DD-Mon-YY, e.g. 27-Sep-16
+                var DMY = columns[0].Split('-');
+
+                // date = 20160927
+                var day = DMY[0].ToInt32();
+                var month = (int)Enum.Parse(typeof(ConvertMonths), DMY[1]);
+                var year = (DMY[2].ToInt32() > 70) ? 1900 + DMY[2].ToInt32() : 2000 + DMY[2].ToInt32();
+                var time = new DateTime(year, month, day, 0, 0, 0);
+
+                // occasionally, the columns will have a '-' instead of a proper value
+                List<Decimal?> ohlc = new List<Decimal?>()
+                {
+                    columns[1] != "-" ? (Decimal?)columns[1].ToDecimal() : null,
+                    columns[2] != "-" ? (Decimal?)columns[2].ToDecimal() : null,
+                    columns[3] != "-" ? (Decimal?)columns[3].ToDecimal() : null,
+                    columns[4] != "-" ? (Decimal?)columns[4].ToDecimal() : null
+                };
+
+                // let's try hard to fix any issues as good as we can
+                // this code assumes that there is at least 1 good value
+                if (ohlc[1] == null) ohlc[1] = ohlc.Where(val => val != null).Max();
+                if (ohlc[2] == null) ohlc[2] = ohlc.Where(val => val != null).Min();
+                if (ohlc[0] == null) ohlc[0] = ohlc.Where(val => val != null).Average();
+                if (ohlc[3] == null) ohlc[3] = ohlc.Where(val => val != null).Average();
```

Comment:
I am not sure what Google does here, I see this about 20 times in 20 years. I agree with your concerns, and it makes me cringe too.
I see the following options:
(1) drop the bar, which will have some side effects.
(2) fill missing values with zero. this might trigger quite erratic reactions in an algorithm.
(3) try to repair. while this is not ideal, I feel that this creates the least amount of side effects

Left code as is, in the spirit of providing the best simulation we can. Basically, we have to admit that downloading data from Google is not ideal. The data from Yahoo are better, but they are not adjusted...